File: everything_parsing/31-lr0-items/js/lr0.mjs
// Module 31 — LR(0) items and the canonical collection (mirror of java/Lr0.java).
//
// item: production + dot.  CLOSURE: dot before nonterminal B -> add B -> . g.
// GOTO(I,X): shift the dot past X, then CLOSURE. Item sets = DFA states; the DFA
// recognizes viable prefixes (Module 30). Complete item -> reduce; outgoing
// terminal edge -> shift; both -> LR(0) conflict.

export class Lr0 {
  constructor(g) {
    this.startAug = g.start + "'";
    this.prods = [{ index: 0, lhs: this.startAug, rhs: [g.start] }];
    for (const p of g.productions) this.prods.push({ index: this.prods.length, lhs: p.lhs, rhs: [...p.rhs] });
    this.nonterminals = new Set([this.startAug, ...g.nonterminals]);
    this.states = [];      // array of Set<"prod.dot">  (we keep the item objects too)
    this.itemLists = [];   // array of item-object arrays, in insertion order
    this.trans = new Map(); // stateIdx -> Map<symbol, stateIdx>
    this._build();
  }

  isNT(s) { return this.nonterminals.has(s); }
  afterDot(it) { const r = this.prods[it.prod].rhs; return it.dot < r.length ? r[it.dot] : null; }

  closure(kernel) {
    const items = [...kernel];
    const seen = new Set(items.map(k));
    let changed = true;
    while (changed) {
      changed = false;
      for (const it of [...items]) {
        const B = this.afterDot(it);
        if (B != null && this.isNT(B)) {
          for (const p of this.prods) {
            if (p.lhs === B) {
              const ni = { prod: p.index, dot: 0 };
              if (!seen.has(k(ni))) { seen.add(k(ni)); items.push(ni); changed = true; }
            }
          }
        }
      }
    }
    return items;
  }

  gotoSet(items, X) {
    const kernel = [];
    for (const it of items)
      if (this.afterDot(it) === X) kernel.push({ prod: it.prod, dot: it.dot + 1 });
    return kernel.length === 0 ? [] : this.closure(kernel);
  }

  _build() {
    const start = this.closure([{ prod: 0, dot: 0 }]);
    this.itemLists.push(start);
    this.states.push(new Set(start.map(k)));
    const index = new Map([[key(start), 0]]);

    const work = [0];
    while (work.length > 0) {
      const si = work.shift();
      const I = this.itemLists[si];

      const symbols = [];
      const sseen = new Set();
      for (const it of I) {
        const s = this.afterDot(it);
        if (s != null && !sseen.has(s)) { sseen.add(s); symbols.push(s); }
      }

      for (const X of symbols) {
        const J = this.gotoSet(I, X);
        if (J.length === 0) continue;
        let target = index.get(key(J));
        if (target === undefined) {
          target = this.itemLists.length;
          this.itemLists.push(J);
          this.states.push(new Set(J.map(k)));
          index.set(key(J), target);
          work.push(target);
        }
        if (!this.trans.has(si)) this.trans.set(si, new Map());
        this.trans.get(si).set(X, target);
      }
    }
  }

  conflicts() {
    const out = [];
    for (let s = 0; s < this.itemLists.length; s++) {
      const complete = [];
      for (const it of this.itemLists[s])
        if (this.afterDot(it) === null && it.prod !== 0) complete.push(it.prod);
      const edges = this.trans.get(s) ?? new Map();
      const shiftSyms = [...edges.keys()].filter(x => !this.isNT(x));

      if (complete.length >= 2)
        out.push({ state: s, kind: "reduce/reduce", detail: "productions [" + complete.join(", ") + "] both reduce here" });
      if (complete.length > 0 && shiftSyms.length > 0)
        out.push({
          state: s, kind: "shift/reduce",
          detail: "reduce by " + this.itemStr({ prod: complete[0], dot: this.prods[complete[0]].rhs.length })
            + " vs shift [" + shiftSyms.join(", ") + "]",
        });
    }
    return out;
  }

  isLR0() { return this.conflicts().length === 0; }

  itemStr(it) {
    const r = this.prods[it.prod].rhs;
    let s = this.prods[it.prod].lhs + " ->";
    for (let i = 0; i <= r.length; i++) {
      if (i === it.dot) s += " .";
      if (i < r.length) s += " " + r[i];
    }
    if (r.length === 0 && it.dot === 0) s += " .";
    return s;
  }

  render() {
    let sb = "augmented grammar:\n";
    for (const p of this.prods)
      sb += "  (" + p.index + ") " + p.lhs + " -> " + (p.rhs.length === 0 ? "epsilon" : p.rhs.join(" ")) + "\n";
    sb += "\n";
    for (let s = 0; s < this.itemLists.length; s++) {
      sb += "I" + s + ":\n";
      for (const it of this.itemLists[s]) sb += "   " + this.itemStr(it) + "\n";
      const edges = this.trans.get(s);
      if (edges && edges.size > 0) {
        const sorted = [...edges.entries()].sort((a, b) => a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0);
        sb += "   ---" + sorted.map(([x, t]) => "  " + x + " -> I" + t).join("") + "\n";
      }
    }
    return sb;
  }
}

const k = it => it.prod + "." + it.dot;
function key(items) { return items.map(k).sort().join(","); }

File: everything_parsing/31-lr0-items/java/Lr0.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;

/**
 * Module 31 — LR(0) items and the canonical collection.
 *
 * An LR(0) item is a production with a dot: E -> E . plus T . The dot marks how
 * much of the RHS has been seen.
 *
 *   CLOSURE(I): while some item  A -> a . B b  is in I and B is a nonterminal,
 *               add  B -> . g  for every production B -> g.
 *   GOTO(I, X): CLOSURE of { A -> a X . b : A -> a . X b in I }.
 *
 * Start from CLOSURE({ S' -> . S }) on the augmented grammar; apply GOTO for
 * every grammar symbol; repeat to a fixed point. The resulting item sets are the
 * STATES of a DFA, and that DFA recognizes exactly the viable prefixes
 * (Module 30).
 *
 * LR(0) parsing decisions come straight from the items: a state with
 * A -> a .  (dot at the end) says "reduce by A -> a"; a state with an outgoing
 * edge on a terminal says "shift". A state with both, or with two complete
 * items, is an LR(0) CONFLICT.
 */
public final class Lr0 {

    /** item = production index into `prods` + dot position (0..rhs.size()). */
    public record Item(int prod, int dot) {}

    public final List<Grammar.Production> prods = new ArrayList<>();   // augmented; prod 0 = S' -> S
    public final String startAug;
    public final Set<String> nonterminals = new LinkedHashSet<>();
    public final List<Set<Item>> states = new ArrayList<>();
    public final Map<Integer, Map<String, Integer>> trans = new LinkedHashMap<>();

    public Lr0(Grammar g) {
        this.startAug = g.start + "'";
        prods.add(new Grammar.Production(0, startAug, List.of(g.start)));
        for (Grammar.Production p : g.productions)
            prods.add(new Grammar.Production(prods.size(), p.lhs(), p.rhs()));
        nonterminals.add(startAug);
        nonterminals.addAll(g.nonterminals);
        build();
    }

    boolean isNT(String s) { return nonterminals.contains(s); }

    List<String> rhsOf(int prod) { return prods.get(prod).rhs(); }
    String lhsOf(int prod) { return prods.get(prod).lhs(); }
    String afterDot(Item it) {
        List<String> r = rhsOf(it.prod());
        return it.dot() < r.size() ? r.get(it.dot()) : null;
    }

    Set<Item> closure(Set<Item> kernel) {
        Set<Item> I = new LinkedHashSet<>(kernel);
        boolean changed = true;
        while (changed) {
            changed = false;
            for (Item it : new ArrayList<>(I)) {
                String B = afterDot(it);
                if (B != null && isNT(B))
                    for (int p = 0; p < prods.size(); p++)
                        if (lhsOf(p).equals(B) && I.add(new Item(p, 0))) changed = true;
            }
        }
        return I;
    }

    Set<Item> gotoSet(Set<Item> I, String X) {
        Set<Item> kernel = new LinkedHashSet<>();
        for (Item it : I)
            if (X.equals(afterDot(it))) kernel.add(new Item(it.prod(), it.dot() + 1));
        return kernel.isEmpty() ? kernel : closure(kernel);
    }

    void build() {
        Set<Item> start = closure(Set.of(new Item(0, 0)));
        states.add(start);
        Map<String, Integer> index = new LinkedHashMap<>();
        index.put(key(start), 0);

        List<Integer> work = new ArrayList<>(List.of(0));
        while (!work.isEmpty()) {
            int si = work.remove(0);
            Set<Item> I = states.get(si);

            Set<String> symbols = new LinkedHashSet<>();
            for (Item it : I) { String s = afterDot(it); if (s != null) symbols.add(s); }

            for (String X : symbols) {
                Set<Item> J = gotoSet(I, X);
                if (J.isEmpty()) continue;
                Integer target = index.get(key(J));
                if (target == null) {
                    target = states.size();
                    states.add(J);
                    index.put(key(J), target);
                    work.add(target);
                }
                trans.computeIfAbsent(si, k -> new TreeMap<>()).put(X, target);
            }
        }
    }

    // ─────────────────────────────────────────── LR(0) analysis

    public record Conflict(int state, String kind, String detail) {}

    public List<Conflict> conflicts() {
        List<Conflict> out = new ArrayList<>();
        for (int s = 0; s < states.size(); s++) {
            List<Integer> complete = new ArrayList<>();
            for (Item it : states.get(s))
                if (afterDot(it) == null && it.prod() != 0) complete.add(it.prod());
            boolean canShift = trans.getOrDefault(s, Map.of()).keySet().stream().anyMatch(x -> !isNT(x));

            if (complete.size() >= 2)
                out.add(new Conflict(s, "reduce/reduce",
                    "productions " + complete + " both reduce here"));
            if (!complete.isEmpty() && canShift)
                out.add(new Conflict(s, "shift/reduce",
                    "reduce by " + itemStr(new Item(complete.get(0), rhsOf(complete.get(0)).size()))
                    + " vs shift " + trans.get(s).keySet().stream().filter(x -> !isNT(x)).toList()));
        }
        return out;
    }

    public boolean isLR0() { return conflicts().isEmpty(); }

    // ─────────────────────────────────────────── rendering

    public String itemStr(Item it) {
        List<String> r = rhsOf(it.prod());
        StringBuilder sb = new StringBuilder(lhsOf(it.prod())).append(" ->");
        for (int i = 0; i <= r.size(); i++) {
            if (i == it.dot()) sb.append(" .");
            if (i < r.size()) sb.append(' ').append(r.get(i));
        }
        if (r.isEmpty() && it.dot() == 0) sb.append(" .");
        return sb.toString();
    }

    public String render() {
        StringBuilder sb = new StringBuilder();
        sb.append("augmented grammar:\n");
        for (Grammar.Production p : prods)
            sb.append("  (").append(p.index()).append(") ").append(p.lhs()).append(" -> ")
              .append(p.rhs().isEmpty() ? "epsilon" : String.join(" ", p.rhs())).append('\n');
        sb.append('\n');
        for (int s = 0; s < states.size(); s++) {
            sb.append("I").append(s).append(":\n");
            for (Item it : states.get(s)) sb.append("   ").append(itemStr(it)).append('\n');
            var t = trans.getOrDefault(s, Map.of());
            if (!t.isEmpty()) {
                sb.append("   ---");
                for (var e : t.entrySet()) sb.append("  ").append(e.getKey()).append(" -> I").append(e.getValue());
                sb.append('\n');
            }
        }
        return sb.toString();
    }

    private String key(Set<Item> I) {
        List<String> ks = new ArrayList<>();
        for (Item it : I) ks.add(it.prod() + "." + it.dot());
        java.util.Collections.sort(ks);
        return String.join(",", ks);
    }
}
